### ISSUE TITLE
[RFC] Try to optimize cache creation

### ISSUE THREAD

**Issue by rousseldenis** [GrandOrgue/grandorgue#1354]
As I try to enhance my understanding of processes behind files management in GO,
I observed recently the process of cache creation.

At first sight, it seems that the configuration and sample files read is quite fast and is parallelized as I saw during that process several read threads monitored with iotop (I'm on Ubuntu).

But, then, the cache creation is quite long and it seems that process uses just one thread.

As I understand write should be unique to be safe, I'm wondering if it exists some possible implementation of 'clever' write ?

@oleg68 If I understand well, the pipe loading is threaded there :

https://github.com/GrandOrgue/grandorgue/blob/master/src/grandorgue/GOOrganController.cpp#L561

And GOCacheWrite called there, iterating between objects for cache creation:

https://github.com/GrandOrgue/grandorgue/blob/master/src/grandorgue/GOOrganController.cpp#L698

Do you think something can be done there ?

**Comment by oleg68** [comment]
I do not have any ideas how to speed the cache creation up. 

The cache file is read and is written sequentially. Of cause, we can use threads for preparing data before writting and for processing data after reading, but the speed will be limited by the disk speed.

Moreover, using the cache was important in case of HDD, but this is not more useful if the sample sets are on SSD, because SSD allows fast random read.

**Comment by rousseldenis** [comment]
> I do not have any ideas how to speed the cache creation up.
> 
> The cache file is read and is written sequentially. Of cause, we can use threads for preparing data before writting and for processing data after reading, but the speed will be limited by the disk speed.
> 
> Moreover, using the cache was important in case of HDD, but this is not more useful if the sample sets are on SSD, because SSD allows fast random read.

The issue I'm facing is that I have it even with a full ssd (even nvme disk with last PCIe x4 capability). For instance, Alessandria sample set is creating the cache during 45 minutes.

But you want to say I maybe need to deactivate cache creation at all?

**Comment by oleg68** [comment]
> But you want to say I maybe need to deactivate cache creation at all?

Yes, I do. I'd suggest to activate cache only if you have HDD.

**Comment by rousseldenis** [comment]
> > But you want to say I maybe need to deactivate cache creation at all?
> 
> Yes, I do. I'd suggest to activate cache only if you have HDD.

Ok, I'll test it.

Even if this subject is still to be investigated

**Comment by jason-barba** [comment]
> Yes, I do. I'd suggest to activate cache only if you have HDD.

How do I deactivate caching?

**Comment by rousseldenis** [comment]
> How do I deactivate caching?

Options > Cache > Automatically manage cache

**Comment by willeke1234** [comment]
Is the cache compressed?

**Comment by rousseldenis** [comment]
> Is the cache compressed?

@willeke1234 if the question is for me: yes

**Comment by rousseldenis** [comment]
> > But you want to say I maybe need to deactivate cache creation at all?
> 
> Yes, I do. I'd suggest to activate cache only if you have HDD.

@oleg68 indeed, deactivating the cache management is resulting in similar loading times than with the cache already created. Nevertheless, those issue is still valid.

**Comment by willeke1234** [comment]
Compression takes a lot of time. On my (old) Mac, HDD, Friesach, update cache:
compressed: 15 minutes
uncompressed: 2.5 minutes

**Comment by kerkovits** [comment]
My experience is that the worst scenario on an SSD is to have a compressed cache. It does not load faster than the original file, and takes much time to create it. On the other hand, having an uncompressed cache on the SSD did result in faster loading times. It seemed that the uncompressed cache was not actually loaded into the RAM, but GO accessed directly the cache on the SSD. Therefore, GO only had to load and parse the ODF, which was very fast, furthermore, I could run memory-intensive tasks parallel to GO this way, as GO did not eat the RAM.

**Comment by rousseldenis** [comment]
> My experience is that the worst scenario on an SSD is to have a compressed cache. It does not load faster than the original file, and takes much time to create it. On the other hand, having an uncompressed cache on the SSD did result in faster loading times. It seemed that the uncompressed cache was not actually loaded into the RAM, but GO accessed directly the cache on the SSD. Therefore, GO only had to load and parse the ODF, which was very fast, furthermore, I could run memory-intensive tasks parallel to GO this way, as GO did not eat the RAM.

Indeed, that was my observations on this weekend.

@oleg68 I will create an issue as errors about cache creation appear at each GO load when cache management is disabled from 3.9.5

**Comment by ahall41** [comment]
You might like to look at pigz (parallel gzip compression) https://zlib.net/pigz/?

Also, one thing I have noticed. If you load your .wav files compressed, GO seems to uncompress and then re-compress when updating the cache. Is it possible to write out the samples in the compressed format (and save the cache uncompressed)?

**Comment by ahall41** [comment]
> Pigz is a command line utility. It is difficult to use it with GrandOrgue. Unfortunately I don't know any implementation of parallel gzip as a c/c++ library.

Source code is also available - you may be able to make use of it?

**Comment by oleg68** [comment]
@ahall41 I think that the business value is quite low but the hardness is quite high. So I'd only add a warning to the "Compress cache" checkbox that compressing the cache takes a lot of time and it is not recommended.

Another useful option is to enable "Loseless compression" and save the samples to the cache "loseless-compressed": in the same form as they are in memory. It would save the disk space about 1.5 times. It is worse than "Compress cache" but it is better than save the cache uncompressed.
